File: user-service/src/main/java/com/github/manosbatsis/services/user/mapper/UserMapper.java
package com.github.manosbatsis.services.user.mapper;

import com.github.manosbatsis.services.user.model.User;
import com.github.manosbatsis.services.user.rest.dto.CreateUserRequest;
import com.github.manosbatsis.services.user.rest.dto.UpdateUserRequest;
import com.github.manosbatsis.services.user.rest.dto.UserResponse;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.NullValuePropertyMappingStrategy;
import org.springframework.context.annotation.Configuration;

@Configuration
@Mapper(
        componentModel = "spring",
        nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE
)
public interface UserMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "deleted", ignore = true)
    User toUser(CreateUserRequest createUserRequest);

    UserResponse toUserResponse(User user);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "deleted", ignore = true)
    @Mapping(target = "email", ignore = true)
    void updateUserFromRequest(UpdateUserRequest updateUserRequest, @MappingTarget User user);
}

File: user-service/src/main/kotlin/com/github/manosbatsis/services/user/mapper/UserMapper.kt
package com.github.manosbatsis.services.user.mapper

import com.github.manosbatsis.services.user.model.User
import com.github.manosbatsis.services.user.rest.dto.CreateUserRequest
import com.github.manosbatsis.services.user.rest.dto.UpdateUserRequest
import com.github.manosbatsis.services.user.rest.dto.UserResponse
import org.mapstruct.MappingTarget

object UserMapper {

    fun instance() = this

    fun toUser(createUserRequest: CreateUserRequest): User {
        val validRequest = createUserRequest.asValidated()
        return User(
            email = validRequest.email,
            fullName = validRequest.fullName,
            address = validRequest.address,
            active = validRequest.active,
        )
    }
    fun toUserResponse(user: User): UserResponse {
        return UserResponse(
            id = user.id,
            email = user.email,
            fullName = user.fullName,
            address = user.address,
            active = user.active,
        )
    }

    fun updateUserFromRequest(updateUserRequest: UpdateUserRequest, @MappingTarget user: User) {
        updateUserRequest.address?.let { user.address = it }
        updateUserRequest.fullName?.let { user.fullName = it }
        updateUserRequest.active?.let { user.active = it }
    }
}
